src/libappimage/desktop_integration/integrator/ResourcesExtractor.cpp

```diff
@@ -0,0 +1,91 @@
+                    }
+
+                    if ((extractIconFiles && isIconFile(filePath))) {
+                        std::vector<char> data(std::istreambuf_iterator<char>(fileItr.read()),
+                                               std::istreambuf_iterator<char>());
+
+                        if (!data.empty())
+                            resources.icons[filePath] = data;
+                    }
+
+                    if ((extractMimeFiles && isMimeFile(filePath))) {
+                        std::vector<char> data(std::istreambuf_iterator<char>(fileItr.read()),
+                                               std::istreambuf_iterator<char>());
+
+                        if (!data.empty())
+                            resources.mimeTypePackages[filePath] = data;
+                    }
+
+                    if ((extractAppDataFile && isAppDataFile(filePath))) {
+                        resources.appStreamPath = filePath;
+                        resources.appStreamData = std::vector<char>(std::istreambuf_iterator<char>(fileItr.read()),
+                                                                    std::istreambuf_iterator<char>());
+                    }
+                }
+
+
+                return resources;
+            }
+
+            bool ResourcesExtractor::isAppDataFile(const std::string& filePath) const {
+                return filePath.find("usr/share/metainfo/") != std::string::npos &&
+                       filePath.find(".appdata.xml") != std::string::npos;
+            }
+
+            bool ResourcesExtractor::isIconFile(const std::string& fileName) const {
+                return (fileName == ".DirIcon") ||
+                       (fileName.find("usr/share/icons") != std::string::npos);
+            }
+
+            bool ResourcesExtractor::isMainDesktopFile(const std::string& fileName) const {
+                return fileName.find(".desktop") != std::string::npos &&
+                       fileName.find('/') == std::string::npos;
+            }
+
+            bool ResourcesExtractor::isMimeFile(const std::string& fileName) const {
+                return fileName.find("usr/share/mime/packages") != std::string::npos &&
+                       fileName.find(".xml") == std::string::npos;
+            }
+
+            void ResourcesExtractor::setExtractDesktopFile(bool extractDesktopFile) {
+                ResourcesExtractor::extractDesktopFile = extractDesktopFile;
+            }
+
+            void ResourcesExtractor::setExtractIconFiles(bool extractIconFiles) {
+                ResourcesExtractor::extractIconFiles = extractIconFiles;
+            }
+
+            void ResourcesExtractor::setExtractAppDataFile(bool extractAppDataFile) {
+                ResourcesExtractor::extractAppDataFile = extractAppDataFile;
+            }
+
+            void ResourcesExtractor::setExtractMimeFiles(bool extractMimeFiles) {
+                ResourcesExtractor::extractMimeFiles = extractMimeFiles;
```

Comment:
I have already defined my positions about using PImpl in private classes. Would add to it that every modification to the class will require the double of work as it should be done in the public interface and in the private.

Can't find a benchmark about `Class::` vs `this->`. It's harmless as it's, let's save a commit for tomorrow.

Reply:
> I have already defined my positions about using PImpl in private classes.

Fair enough.

> Would add to it that every modification to the class will require the double of work as it should be done in the public interface and in the private.

Why should it? The point is, every class should define a public interface only, PImpl makes things a lot easier as the data can be managed next to the actual implementation of the methods. You don't have to hide/extract methods (only private ones), everything that _can_ be in the public method definition can just remain there.

It's just a matter of style and correct design (see also design-by-contract). Also, it'll save unnecessary recompilation whenever the header needs to change, and that's something that I really dislike about C++ (having to re-compile all units when you change private fields in a header).

> Would add to it that every modification to the class will require the double of work as it should be done in the public interface and in the private.

Was more of a general remark. Of course this will work, too.